api/src/main/java/org/apache/iceberg/expressions/Predicate.java

```diff
@@ -19,15 +19,44 @@
 
 package org.apache.iceberg.expressions;
 
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import java.util.Collection;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
 public abstract class Predicate<T, R extends Reference> implements Expression {
   private final Operation op;
   private final R ref;
   private final Literal<T> literal;
+  private final Set<Literal<T>> literalSet;
```

Comment:
I'm not sure about the choice to use the same `Predicate` with either a set or a literal. I think this should probably be two separate classes, at least in the case of a bound predicate.

Reply:
Thanks for the review!
Yeah, I agree that it is better to have two classes for bound predicate as it will only be generated as expression during binding. 
For `UnboundPredicate` and `Predicate`, I still think it might be good to be a single class. For example, `UnboundPredicate` for `IN` predicate with a single item should return an `EQ` predicate. The separation will complicate it.
Please let me know your comments. Thanks.

Reply:
> UnboundPredicate for IN predicate with a single item should return an EQ predicate. The separation will complicate it.

I disagree because that over-complicates the predicate classes. The factory methods in `Expressions` are allowed to change the predicate that is returned; for example, `and(true, p)` returns just `p`. That's what we should use to return an equality predicate. The predicate class itself should not include logic for changing itself.

Reply:
@rdblue Sorry for the misunderstanding.

In this change, only factory methods in `Expressions` and the `bind` in `UnboundPredicate` can change the returned predicate.

What I meant is that the factory method `in` will return an `EQ` predicate if only a single element is provided because UnboundPredicate factory method for IN operation with a single item should return `EQ` predicate.

As the `bind` returns `Expression`, it is straightforward to separated the `BoundPredicate`.

But in the case of the factory method, if it also returns `Expression` or a new interface class instead of `UnboundPredicate`, then later when it is used, it has to be explicitly cast by checking the `operation`. This seems over-complicated compared to the current approach.